src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1972,6 +1972,39 @@ private void ImportRefAnyVal(int token)
 
         private void ImportCkFinite()
         {
+            StackEntry value = _stack.Pop();
+
+            var targetType = GetLLVMTypeForTypeDesc(value.Type);
+            var width = LLVM.SizeOfTypeInBits(default, targetType);
+            LLVMValueRef exponentMask;
+            LLVMTypeRef intTypeRef;
+
+            if (width == 32)
+            {
+                intTypeRef = LLVM.Int32Type();
+                exponentMask = LLVM.ConstInt(intTypeRef, 0x7F800000, LLVMMisc.False);
+            }
+            else
+            {
+                intTypeRef = LLVM.Int64Type();
+                exponentMask = LLVM.ConstInt(intTypeRef, 0x7FF0000000000000, LLVMMisc.False);
+            }
+
+            var valRef = value.ValueAsType(intTypeRef, _builder);
```

Comment:
@morganbr, if this line is replaced with a constant: `var valRef = LLVM.ConstInt(intTypeRef, 0x7F7FFFC0, LLVMMisc.False);`, it passes the check for all values until `0xFF7FFFFF` and fails `0x7F800000`. That is the expected behavior for 32-bit.

It seems like there is some kind of shifting going on with `ValueAsType` that it fails anything greater than `0x7F7FFFBF`. Is there anyway to debug the runtime value in decimal or binary? Maybe by emitting the `.ll` file from emscripten in the  middle of the process?

Reply:
ValueAsType might be doing sign extension or somehow messing up the value based on what it thinks the type system is telling it. Do you know what the LLVM opcodes are for the failing bits? I don't think i have anything special setup and I get a text file with all of the LLVM ops in HelloWasm\obj\Debug\wasm\native\HelloWasm.txt

Reply:
This `value.ValueAsType` calls `LoadExpression.ValueAsTypeInternal` and eventual LLVM ops are:

```llvm
%CastIfNecessaryPtr = bitcast i8* %2 to float*
%ldvalue = load float, float* %CastIfNecessaryPtr
%fptoui = fptoui float %ldvalue to i32
%and = and i32 %fptoui, 2139095040
```

Should it be `fptosi`?

I tried explicitly calling:

```c#
var valRef = LLVM.BuildPointerCast(_builder, ((LoadExpressionEntry)value).RawLLVMValue, intTypeRef, "CastIfNecessaryPtr");
```

which emits:

```llvm
%CastIfNecessaryPtr = ptrtoint i8* %2 to i32
%and = and i32 %CastIfNecessaryPtr, 2139095040
```

With that it passes all the values, false positive, even the `0x7F800000` which should trap.

There are some TODOs related to signed cast in EvaluationStack, could be related to that.